### PR TITLE
Resolve `LogsSuite` errors

### LINKED ISSUE #8780: Resolve `LogsSuite` errors
### Problem

We have the following tests suite `LogsSuite` that has problems when it's running against modularization codebase.

### Solution

Resolve issues with the test suite. When ready add @hapitest annotations to the passing tests.

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by natanasow**
**Description**:
 - Closes #8780 
 - Adds a transaction/query-scoped `TinybarsValues` object to the root `MessageFrame` context; this allows the `GasCalculator` to remain a singleton, while still customizing the `log0-4` gas calculations according to the relative Hedera `rbh` and `gas` prices.

### CODE REVIEW

**Review by tinker-michaelj — CHANGES_REQUESTED** [review]
Hi @natanasow, I believe the correct fix here is to finish implementation of [the `CustomGasCalculator`](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/CustomGasCalculator.java#L31) so that it behaves identically to the mono-service [`GasCalculatorHederaV22`](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/gascalculator/GasCalculatorHederaV22.java).

A lot of the mono-service code will come over directly. But for `LOG` operation gas costs in particular, the `GasCalculatorHederaV22` does need several pieces of Hedera context:
 1. The [value of the `cache.records.ttl` property](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/gascalculator/GasCalculatorHederaV19.java#L88).
 2. The [type of top-level contract operation](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/gascalculator/GasCalculatorHederaUtil.java#L85) we are executing (`ContractCall`, `ContractCreate`, or `EthereumTransaction`).
 3. [The current exchange rate (from tinycents to tinybars)](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/gascalculator/GasCalculatorHederaUtil.java#L70).
 4. [The current `rbh` resource price for each contract transaction type.](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/gascalculator/GasCalculatorHederaUtil.java#L62)

Note that all of these values can, in principle, change from transaction to transaction.

So it is not really "natural" for `CustomGasCalculator` to have the `@Singleton` scope; the more natural scopes are [`@TransactionScope`](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/TransactionComponent.java#L27)  and [`@QueryScope`](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/QueryComponent.java#L27) , so we create a new `GasCalculator` for each transaction and query. (This also protects us from race conditions between query threads and the handle thread.)

When we switch the `CustomGasCalculator` to these scopes, we can get,
  1. The `cache.records.ttl` property value from the `MessageFrame` as we do any other property value (e.g. [here](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/v034/Version034FeatureFlags.java#L39)).
  2. The top-level type by adding a `@BindsInstance HederaFunctionality functionality` to the  [`Factory.create()` methods](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/TransactionComponent.java#L31) of the components (and then e.g. passing in `HederaFunctionality.ContractCreate` from the `ContractCreateHandler`, etc.)
  3. The current exchange rate from the `MessageFrame`'s updater as is done [here](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/ExchangeRateSystemContract.java#L62).
  4. _(Change in `hedera-base`)_ The current price of the `rbh` resource by adding a method to `HandleContext` that exposes [this `FeeManager` method](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/FeeManager.java#L183) (should probably be named `getResourcePrices()` instead of `getFeeData()` actually).

Does that make sense?

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/FrameRunner.java [hunk 1]
```diff
@@ -125,6 +125,8 @@ private long effectiveGasUsed(final long gasLimit, @NonNull final MessageFrame f
                 * Math.min(frame.getSelfDestructs().size(), nominalUsed / gasCalculator.getMaxRefundQuotient());
         nominalUsed -= (selfDestructRefund + frame.getGasRefund());
         final var maxRefundPercent = contractsConfigOf(frame).maxRefundPercentOfGasLimit();
+        // include the base gas consumption of 21_000 in the final calculation
+        nominalUsed = nominalUsed < 21_000L && nominalUsed > 0 ? nominalUsed + 21_000L : nominalUsed;
```
Comment (netopyr) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/FrameRunner.java:129]: Please use a constant instant of a magic number

Comment (natanasow) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/FrameRunner.java:129]: This piece of code will be completely removed regarding the @tinker-michaelj comment above.

**Comment by tinker-michaelj** [comment]
> > Hi @natanasow, I believe the correct fix here is to finish implementation of [the `CustomGasCalculator`](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/CustomGasCalculator.java#L31) so that it behaves identically to the mono-service [`GasCalculatorHederaV22`](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/gascalculator/GasCalculatorHederaV22.java).
> > A lot of the mono-service code will come over directly. But for `LOG` operation gas costs in particular, the `GasCalculatorHederaV22` does need several pieces of Hedera context:
> > ```
> > 1. The [value of the `cache.records.ttl` property](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/gascalculator/GasCalculatorHederaV19.java#L88).
> > 
> > 2. The [type of top-level contract operation](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/gascalculator/GasCalculatorHederaUtil.java#L85) we are executing (`ContractCall`, `ContractCreate`, or `EthereumTransaction`).
> > 
> > 3. [The current exchange rate (from tinycents to tinybars)](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/gascalculator/GasCalculatorHederaUtil.java#L70).
> > 
> > 4. [The current `rbh` resource price for each contract transaction type.](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/gascalculator/GasCalculatorHederaUtil.java#L62)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Note that all of these values can, in principle, change from transaction to transaction.
> > So it is not really "natural" for `CustomGasCalculator` to have the `@Singleton` scope; the more natural scopes are [`@TransactionScope`](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/TransactionComponent.java#L27) and [`@QueryScope`](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/QueryComponent.java#L27) , so we create a new `GasCalculator` for each transaction and query. (This also protects us from race conditions between query threads and the handle thread.)
> > When we switch the `CustomGasCalculator` to these scopes, we can get,
> > ```
> > 1. The `cache.records.ttl` property value from the `MessageFrame` as we do any other property value (e.g. [here](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/v034/Version034FeatureFlags.java#L39)).
> > 
> > 2. The top-level type by adding a `@BindsInstance HederaFunctionality functionality` to the  [`Factory.create()` methods](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/TransactionComponent.java#L31) of the components (and then e.g. passing in `HederaFunctionality.ContractCreate` from the `ContractCreateHandler`, etc.)
> > 
> > 3. The current exchange rate from the `MessageFrame`'s updater as is done [here](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/ExchangeRateSystemContract.java#L62).
> > 
> > 4. _(Change in `hedera-base`)_ The current price of the `rbh` resource by adding a method to `HandleContext` that exposes [this `FeeManager` method](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/FeeManager.java#L183) (should probably be named `getResourcePrices()` instead of `getFeeData()` actually).
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Does that make sense?
> 
> Hi @tinker-michaelj, thank you very much for the detailed instructions on how to implement the GasCalculator but I have some questions 🙏 .
> 
> I'm not very familiar with the scope concept. I tried several things but with no success. What do you mean by "...so we create a new GasCalculator for each transaction and query..."? I have to delete the current `CustomGasCalculator` and create new ones - `TransactionGasCalculator` and `QueryGasCalculator` which will extend `LondonGasCalculator`? After that, I should add them in both `QueryModule.java` and `TransactionModule.java` as following lines and remove `bindGasCalculator` from `ContractServiceModule`?
> 
> ```java
> // QueryModule.java
> @Provides
> @QueryScope
> static QueryGasCalculator provideGasCalculator() {
>     return new QueryGasCalculator();
> }
> ```
> 
> ```java
> // TransactionModule.java
> @Provides
> @TransactionScope
> static TransactionGasCalculator provideGasCalculator() {
>     return new TransactionGasCalculator();
> }
> ```
> 
> If I do that I get this error
> 
> ```
> [Dagger/MissingBinding] org.hyperledger.besu.evm.gascalculator.GasCalculator cannot be provided without an @Provides-annotated method.
> org.hyperledger.besu.evm.gascalculator.GasCalculator is injected at
> com.hedera.node.app.service.contract.impl.infra.HevmTransactionFactory(…, gasCalculator, …)
> com.hedera.node.app.service.contract.impl.infra.HevmTransactionFactory is injected at
> com.hedera.node.app.service.contract.impl.exec.ContextTransactionProcessor(…, hevmTransactionFactory, …)
> com.hedera.node.app.service.contract.impl.exec.ContextTransactionProcessor is requested at
> com.hedera.node.app.service.contract.impl.exec.TransactionComponent.contextTransactionProcessor() [com.hedera.node.app.service.contract.impl.ContractServiceComponent → com.hedera.node.app.service.contract.impl.exec.TransactionComponent]
> It is also requested at:
> com.hedera.node.app.service.contract.impl.exec.gas.CustomGasCharging(gasCalculator)
> com.hedera.node.app.service.contract.impl.exec.systemcontracts.ExchangeRateSystemContract(gasCalculator)
> com.hedera.node.app.service.contract.impl.exec.systemcontracts.HtsSystemContract(gasCalculator, …)
> com.hedera.node.app.service.contract.impl.exec.systemcontracts.PrngSystemContract(gasCalculator)
> com.hedera.node.app.service.contract.impl.exec.v030.V030Module.provideStaticCallOperation(gasCalculator, …)
> com.hedera.node.app.service.contract.impl.exec.v030.V030Module.provideExtCodeCopyOperation(gasCalculator, …)
> com.hedera.node.app.service.contract.impl.exec.v030.V030Module.providePrecompileContractRegistry(gasCalculator)
> com.hedera.node.app.service.contract.impl.exec.v030.V030Module.provideCallOperation(gasCalculator, …)
> com.hedera.node.app.service.contract.impl.exec.v030.V030Module.provideLog4Operation(gasCalculator)
> com.hedera.node.app.service.contract.impl.exec.v030.V030Module.provideSLoadOperation(gasCalculator, …)
> and 64 others
> The following other entry points also depend on it:
> com.hedera.node.app.service.contract.impl.exec.QueryComponent.contextQueryProcessor() [com.hedera.node.app.service.contract.impl.ContractServiceComponent → com.hedera.node.app.service.contract.impl.exec.QueryComponent]
> ```
> 
> I guess, I'm doing something wrong 😕 Could you give a little bit more information about scopes or point me out already existing classes that are defined in both Transaction and Query scope for reference/example?
> 
> PS: Aside from separating the `CustomGasCalculator` into two scopes, I think I got how to get the needed props for `logOperationGasCost`.

Hi @natanasow, we will only have one `CustomGasCalculator` implementation; but instances of that implementation will be provided in both the `TransactionModule` _and_ the `QueryModule`. (This is similar to how both modules provide a `HederaWorldUpdater.Enhancement` instance.)

(This is the sense in which we create a new `CustomGasCalculator` for each query and transaction we process. E.g., each  transaction is processed with a `TransactionComponent`; and each time Dagger2 constructs a `TransactionComponent` it creates instances of all types provided in the `TransactionModule`.)

So if you have worked out how to provide your `CustomGasCalculator` with those bits of context in both the `QueryModule` and `TransactionModule`, I think you're pretty much there! 

You are correct we can now remove `bindGasCalculator()` from `ContractServiceModule`; that module creates `@Singleton`-scoped objects that are created on starting up the app and re-used for every transaction and query.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/TransactionModule.java [hunk 2]
```diff
@@ -53,6 +55,12 @@
 
 @Module(includes = {TransactionConfigModule.class, TransactionInitialStateModule.class})
 public interface TransactionModule {
+    @Provides
+    @TransactionScope
+    static CustomGasCalculator provideCustomGasCalculator(@NonNull HederaFunctionality functionality) {
+        return new CustomGasCalculator();
```
Comment (natanasow) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/TransactionModule.java:61]: @tinker-michaelj 

Would you take a look at the latest commit? I've added `CustomGasCalculator` to `Transaction` and `Query` scopes. I was thinking about how to inject `Query/Handle` context or `HederaFunctionality functionality` from Component to `new CustomGasCalculator()`. I have no success with passing `functionality` to the `CustomGasCalculator` constructor :confused:. 

Do you have something in mind about how to use `context` or `functionality` in `CustomGasCalculatorl.logOperationGasCost` method?

Comment (tinker-michaelj) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/TransactionModule.java:61]: Hi @natanasow, do you mind if I open a PR to this branch? 

Or maybe we need an out-of-band call; some of these details are definitely hard to nail down asynchronously via GitHub. 😇

Comment (natanasow) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/TransactionModule.java:61]: @tinker-michaelj feel free to open a completely new branch and or commit whatever you want to this one.

**Review by tinker-michaelj — APPROVED** [review]
LGTM

**Pull request merged**